File: src/core/extended_send_manager.py
import os
import json
from PySide6.QtCore import QObject, Signal, QTimer


class ExtendedSendManager(QObject):
    """扩展发送管理类，负责管理多条数据的发送"""
    
    send_started = Signal()           # 发送开始信号
    send_finished = Signal()          # 发送完成信号
    send_progress = Signal(int, int)  # 发送进度信号 (当前序号, 总数)
    data_sent = Signal(bytes)         # 数据已发送信号
    error_occurred = Signal(str)      # 错误信号
    items_changed = Signal()          # 数据列表改变信号
    
    def __init__(self, send_func, parent=None):
        super().__init__(parent)
        self.send_func = send_func  # 注入的发送函数
        self.items = []  # 扩展发送数据列表
        self.is_sending = False
        self.is_looping = False
        self.current_index = 0
        self.send_timer = QTimer()
        self.send_timer.setSingleShot(True)
        self.send_timer.timeout.connect(self._on_send_timer_timeout)
        
        # 配置文件路径
        self.config_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "config")
        self.config_file = os.path.join(self.config_dir, "extended_send.json")
        
        # 加载配置
        self._load_config()
    
    def _load_config(self):
        """加载配置"""
        try:
            if os.path.exists(self.config_file):
                with open(self.config_file, 'r', encoding='utf-8') as f:
                    config_data = json.load(f)
                    self.items = config_data.get('items', [])
                    self._reorder_items()
        except Exception as e:
            print(f"加载扩展发送配置失败: {e}")
    
    def _save_config(self):
        """保存配置"""
        try:
            os.makedirs(self.config_dir, exist_ok=True)
            config_data = self.save_to_config()
            with open(self.config_file, 'w', encoding='utf-8') as f:
                json.dump(config_data, f, ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"保存扩展发送配置失败: {e}")
    
    def add_item(self, data, is_hex=True, comment="", delay=1000):
        """添加一条数据"""
        item = {
            'id': self._generate_id(),
            'data': data,
            'is_hex': is_hex,
            'comment': comment,
            'delay': delay,
            'sort_order': 0
        }
        self.items.append(item)
        self._save_config()
        self.items_changed.emit()
        return item
    
    def remove_item(self, item_id):
        """删除一条数据"""
        self.items = [item for item in self.items if item['id'] != item_id]
        self._reorder_items()
        self._save_config()
        self.items_changed.emit()
    
    def update_item(self, item_id, **kwargs):
        """更新一条数据"""
        for item in self.items:
            if item['id'] == item_id:
                item.update(kwargs)
                self._save_config()
                self.items_changed.emit()
                break
    
    def move_item(self, item_id, direction):
        """移动数据顺序 (direction: -1=上移, 1=下移)"""
        index = next((i for i, item in enumerate(self.items) if item['id'] == item_id), None)
        if index is None:
            return
        
        new_index = index + direction
        if 0 <= new_index < len(self.items):
            self.items[index], self.items[new_index] = self.items[new_index], self.items[index]
            self._reorder_items()
            self._save_config()
            self.items_changed.emit()
    
    def _reorder_items(self):
        """重新排序"""
        for i, item in enumerate(self.items):
            item['sort_order'] = i
    
    def _generate_id(self):
        """生成唯一ID"""
        if not self.items:
            return 1
        return max(item['id'] for item in self.items) + 1
    
    def get_sorted_items(self):
        """获取序号大于0的数据项，按顺序排列"""
        return sorted(
            [item for item in self.items if item.get('sort_order', 0) > 0],
            key=lambda x: x['sort_order']
        )
    
    def clear_items(self):
        """清空所有数据"""
        self.items.clear()
        self._save_config()
        self.items_changed.emit()
    
    def send_single(self, item_id):
        """发送单条数据"""
        item = next((item for item in self.items if item['id'] == item_id), None)
        if item:
            self._send_item(item)
    
    def send_multiple(self, loop=False):
        """发送多条数据"""
        if self.is_sending:
            return
        
        self.is_sending = True
        self.is_looping = loop
        self.current_index = 0
        self.send_started.emit()
        
        sorted_items = self.get_sorted_items()
        if not sorted_items:
            self.is_sending = False
            self.send_finished.emit()
            return
        
        self._send_next_item(sorted_items)
    
    def _send_next_item(self, items):
        """发送下一条数据"""
        if not self.is_sending:
            return
        
        if self.current_index >= len(items):
            if self.is_looping:
                self.current_index = 0
            else:
                self.is_sending = False
                self.send_finished.emit()
                return
        
        item = items[self.current_index]
        success = self._send_item(item)
        
        if not success:
            self.is_sending = False
            self.send_finished.emit()
            return
        
        self.send_progress.emit(item['id'], len(items))
        
        self.current_index += 1
        
        # 延时后发送下一条
        if item['delay'] > 0:
            self.send_timer.start(item['delay'])
        else:
            self._send_next_item(items)
    
    def _on_send_timer_timeout(self):
        """发送延时定时器超时"""
        sorted_items = self.get_sorted_items()
        self._send_next_item(sorted_items)
    
    def _send_item(self, item):
        """发送单条数据，返回是否成功"""
        try:
            data = item['data']
            is_hex = item['is_hex']
            
            if not data:
                self.error_occurred.emit("数据为空")
                return False
            
            if is_hex:
                hex_str = data.replace(' ', '')
                send_data = bytes.fromhex(hex_str)
            else:
                send_data = data.encode('utf-8')
            
            # 调用注入的发送函数
            success = self.send_func(send_data)
            
            if not success:
                self.error_occurred.emit("发送失败")
                return False
            
            self.data_sent.emit(send_data)
            return True
        except Exception as e:
            self.error_occurred.emit(f"发送失败: {str(e)}")
            return False
    
    def stop_sending(self):
        """停止发送"""
        self.is_sending = False
        self.is_looping = False
        self.send_timer.stop()
    
    def load_from_config(self, config_data):
        """从配置数据加载"""
        self.items = config_data.get('items', [])
        self._reorder_items()
        self._save_config()
        self.items_changed.emit()
    
    def save_to_config(self):
        """保存为配置数据"""
        return {
            'items': self.items,
            'settings': {
                'loop_send': self.is_looping,
                'multi_send': True,
                'default_delay': 1000
            }
        }
    
    def import_from_file(self, file_path):
        """从文件导入配置"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                config_data = json.load(f)
            self.load_from_config(config_data)
            return True
        except Exception as e:
            self.error_occurred.emit(f"导入失败: {str(e)}")
            return False
    
    def export_to_file(self, file_path):
        """导出配置到文件"""
        try:
            config_data = self.save_to_config()
            with open(file_path, 'w', encoding='utf-8') as f:
                json.dump(config_data, f, ensure_ascii=False, indent=2)
            return True
        except Exception as e:
            self.error_occurred.emit(f"导出失败: {str(e)}")
            return False

File: src/core/extended_send_manager.cpp
#include "extended_send_manager.h"

#include <QFile>
#include <QJsonDocument>
#include <QDir>

ExtendedSendManager::ExtendedSendManager(SendFunc sendFunc, const QString &configDir,
                                         QObject *parent)
    : QObject(parent), m_sendFunc(std::move(sendFunc))
{
    QDir dir(configDir);
    if (!dir.exists()) {
        dir.mkpath(".");
    }
    m_configPath = dir.absoluteFilePath("extended_send.json");

    m_debounceTimer = new QTimer(this);
    m_debounceTimer->setSingleShot(true);
    m_debounceTimer->setInterval(500);
    connect(m_debounceTimer, &QTimer::timeout, this, &ExtendedSendManager::saveItems);

    loadItems();
}

int ExtendedSendManager::generateId() {
    static int counter = 0;
    return ++counter;
}

void ExtendedSendManager::loadItems() {
    QFile file(m_configPath);
    if (!file.open(QIODevice::ReadOnly)) return;

    QJsonDocument doc = QJsonDocument::fromJson(file.readAll());
    if (!doc.isObject()) return;

    QJsonObject root = doc.object();
    QJsonArray items = root.value("items").toArray();

    int maxId = 0;
    m_items.clear();
    for (const auto &v : items) {
        QJsonObject obj = v.toObject();
        SendItem item;
        item.id = obj.value("id").toInt();
        item.data = obj.value("data").toString();
        item.isHex = obj.value("is_hex").toBool();
        item.comment = obj.value("comment").toString();
        item.delay = obj.value("delay").toInt(100);
        item.sortOrder = obj.value("sort_order").toInt(0);
        m_items.append(item);
        if (item.id > maxId) maxId = item.id;
    }

    // Update static counter
    static int *counterPtr = nullptr;
    if (!counterPtr) {
        static int counter = 0;
        counterPtr = &counter;
    }
    *counterPtr = qMax(*counterPtr, maxId);
}

void ExtendedSendManager::saveItems() {
    QJsonArray items;
    for (const auto &item : m_items) {
        QJsonObject obj;
        obj["id"] = item.id;
        obj["data"] = item.data;
        obj["is_hex"] = item.isHex;
        obj["comment"] = item.comment;
        obj["delay"] = item.delay;
        obj["sort_order"] = item.sortOrder;
        items.append(obj);
    }

    QJsonObject root;
    root["items"] = items;

    QFile file(m_configPath);
    if (file.open(QIODevice::WriteOnly)) {
        file.write(QJsonDocument(root).toJson());
    }
}

void ExtendedSendManager::scheduleSave() {
    m_debounceTimer->start();
}

void ExtendedSendManager::flush() {
    if (m_debounceTimer->isActive()) {
        m_debounceTimer->stop();
        saveItems();
    }
}

int ExtendedSendManager::addItem(const QString &data, bool isHex,
                                  const QString &comment, int delay) {
    SendItem item;
    item.id = generateId();
    item.data = data;
    item.isHex = isHex;
    item.comment = comment;
    item.delay = delay;
    item.sortOrder = m_items.size() + 1;

    m_items.append(item);
    emit itemsChanged();
    scheduleSave();
    return item.id;
}

void ExtendedSendManager::removeItem(int itemId) {
    for (int i = 0; i < m_items.size(); ++i) {
        if (m_items[i].id == itemId) {
            m_items.removeAt(i);
            emit itemsChanged();
            scheduleSave();
            return;
        }
    }
}

void ExtendedSendManager::updateItem(int itemId, const QString &key, const QVariant &value) {
    for (auto &item : m_items) {
        if (item.id == itemId) {
            if (key == "data") item.data = value.toString();
            else if (key == "is_hex") item.isHex = value.toBool();
            else if (key == "comment") item.comment = value.toString();
            else if (key == "delay") item.delay = value.toInt();
            else if (key == "sort_order") item.sortOrder = value.toInt();

            emit itemsChanged();
            scheduleSave();
            return;
        }
    }
}

void ExtendedSendManager::moveItem(int itemId, int direction) {
    for (int i = 0; i < m_items.size(); ++i) {
        if (m_items[i].id == itemId) {
            int newIdx = i + direction;
            if (newIdx >= 0 && newIdx < m_items.size()) {
                m_items.swapItemsAt(i, newIdx);
                // Update sort orders
                for (int j = 0; j < m_items.size(); ++j) {
                    m_items[j].sortOrder = j + 1;
                }
                emit itemsChanged();
                scheduleSave();
            }
            return;
        }
    }
}

void ExtendedSendManager::clearItems() {
    m_items.clear();
    emit itemsChanged();
    scheduleSave();
}

QList<SendItem> ExtendedSendManager::getSortedItems() const {
    QList<SendItem> sorted;
    for (const auto &item : m_items) {
        if (item.sortOrder > 0) {
            sorted.append(item);
        }
    }
    std::sort(sorted.begin(), sorted.end(), [](const SendItem &a, const SendItem &b) {
        return a.sortOrder < b.sortOrder;
    });
    return sorted;
}

void ExtendedSendManager::sendSingle(int itemId) {
    for (const auto &item : m_items) {
        if (item.id == itemId) {
            sendItem(item);
            return;
        }
    }
}

void ExtendedSendManager::sendMultiple(bool loop) {
    if (m_sending) return;

    m_sendingItems = getSortedItems();
    if (m_sendingItems.isEmpty()) {
        emit errorOccurred("No items to send");
        return;
    }

    m_sending = true;
    m_loopSend = loop;
    m_sendIndex = 0;
    emit sendStarted();
    sendNextItem();
}

void ExtendedSendManager::sendNextItem() {
    if (!m_sending || m_sendIndex >= m_sendingItems.size()) {
        if (m_loopSend && m_sending) {
            // Loop: restart
            m_sendIndex = 0;
            emit sendProgress(0, m_sendingItems.size());
            sendNextItem();
            return;
        }
        // Done
        m_sending = false;
        emit sendFinished();
        return;
    }

    emit sendProgress(m_sendIndex + 1, m_sendingItems.size());

    const SendItem &item = m_sendingItems[m_sendIndex];
    int delay = (m_sendIndex == 0) ? 0 : item.delay;

    QTimer::singleShot(delay, this, [this, item]() {
        if (!m_sending) return;
        sendItem(item);
        m_sendIndex++;
        sendNextItem();
    });
}

void ExtendedSendManager::stopSending() {
    m_sending = false;
    m_loopSend = false;
    emit sendFinished();
}

bool ExtendedSendManager::sendItem(const SendItem &item) {
    QByteArray data;
    if (item.isHex) {
        // Validate hex string
        QString cleaned = item.data;
        cleaned.remove(' ');
        if (cleaned.size() % 2 != 0) {
            emit errorOccurred("Invalid hex data (odd length): " + item.data);
            return false;
        }
        for (const QChar &ch : cleaned) {
            if (!ch.isLetterOrNumber()) {
                emit errorOccurred("Invalid hex data: " + item.data);
                return false;
            }
        }
        data = QByteArray::fromHex(cleaned.toLatin1());
    } else {
        // Decode ASCII escapes
        QString decoded = item.data;
        decoded.replace("\\r", "\r");
        decoded.replace("\\n", "\n");
        decoded.replace("\\t", "\t");
        decoded.replace("\\0", "\0");
        decoded.replace("\\\\", "\\");
        data = decoded.toUtf8();
    }

    if (m_sendFunc) {
        bool ok = m_sendFunc(data);
        if (ok) {
            emit dataSent(data);
        }
        return ok;
    }
    return false;
}

void ExtendedSendManager::importFromFile(const QString &filePath) {
    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly)) {
        emit errorOccurred("Cannot open file: " + filePath);
        return;
    }

    QJsonDocument doc = QJsonDocument::fromJson(file.readAll());
    if (!doc.isObject()) {
        emit errorOccurred("Invalid JSON format");
        return;
    }

    QJsonObject root = doc.object();
    QJsonArray items = root.value("items").toArray();

    m_items.clear();
    for (const auto &v : items) {
        QJsonObject obj = v.toObject();
        SendItem item;
        item.id = obj.value("id").toInt();
        item.data = obj.value("data").toString();
        item.isHex = obj.value("is_hex").toBool();
        item.comment = obj.value("comment").toString();
        item.delay = obj.value("delay").toInt(100);
        item.sortOrder = obj.value("sort_order").toInt(0);
        m_items.append(item);
    }

    emit itemsChanged();
    saveItems();
}

void ExtendedSendManager::exportToFile(const QString &filePath) const {
    QJsonArray items;
    for (const auto &item : m_items) {
        QJsonObject obj;
        obj["id"] = item.id;
        obj["data"] = item.data;
        obj["is_hex"] = item.isHex;
        obj["comment"] = item.comment;
        obj["delay"] = item.delay;
        obj["sort_order"] = item.sortOrder;
        items.append(obj);
    }

    QJsonObject root;
    root["items"] = items;

    QFile file(filePath);
    if (file.open(QIODevice::WriteOnly)) {
        file.write(QJsonDocument(root).toJson());
    }
}
